src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -498,6 +585,44 @@ public DynamicObject getEncoding(int index) {
     }
 
     @NodeChildren({ @NodeChild("first"), @NodeChild("second") })
+    public static abstract class CheckRopeEncodingNode extends RubyNode {
+
+        @Child private NegotiateCompatibleRopeEncodingNode negotiateCompatibleEncodingNode;
+        @Child private ToEncodingNode toEncodingNode;
+
+        public static CheckRopeEncodingNode create() {
+            return EncodingNodesFactory.CheckRopeEncodingNodeGen.create(null, null);
+        }
+
+        public CheckRopeEncodingNode() {
+            negotiateCompatibleEncodingNode = EncodingNodesFactory.NegotiateCompatibleRopeEncodingNodeGen.create(null, null);
+        }
+
+        public abstract Encoding executeCheckEncoding(Rope first, Rope second);
+
+        @Specialization
+        public Encoding checkEncoding(Rope first, Rope second,
+                @Cached("create()") BranchProfile errorProfile) {
+            final Encoding negotiatedEncoding = negotiateCompatibleEncodingNode.executeNegotiate(first, second);
+
+            if (negotiatedEncoding == null) {
+                errorProfile.enter();
+
+                if (toEncodingNode == null) {
+                    CompilerDirectives.transferToInterpreterAndInvalidate();
+                    toEncodingNode = insert(ToEncodingNode.create());
+                }
```

Comment:
Please extract the lazy-node pattern in its own method, including the executeToEncoding.